### PR TITLE
Add some unit test (workflow) + DX improvements

### PR DESCRIPTION
**Pull request by dawehner**

### CODE REVIEW

**Comment by dawehner** [comment]
So +1 or -1?

**Comment by bojanz** [comment]
Working on tweaking this a bit.

File: src/Plugin/Workflow/Workflow.php [hunk 1]
```diff
@@ -55,18 +55,30 @@ public function __construct(array $configuration, $plugin_id, $plugin_definition
...
+
     // Populate value objects for states and transitions.
     foreach ($plugin_definition['states'] as $id => $state_definition) {
-      $this->states[$id] = new WorkflowState($id, $state_definition['label']);
+      if ($state_definition instanceof WorkflowState) {
+        $this->states[$id] = $state_definition;
+      }
+      else {
+        $this->states[$id] = new WorkflowState($id, $state_definition['label']);
+      }
```
Comment (LionsAd) [inline src/Plugin/Workflow/Workflow.php]: I agree, this smells code smelly ...

I personally am okay with passing a well defined array-of-doom, but its not the season for that anymore ...

**Comment by bojanz** [comment]
Committed a version without the value objects in 1f8cac99802f2e25ac7616d62b8de9eb17cb904c.

As I said on IRC, it feels weird for the same constructor to support two different ways of creating the object. Either we do array-of-doom or we create the value objects in the WorkflowManager / static factory. Plugins in general are very influenced by where they get the data from, so this piece of ugliness is inherited.

**Pull request closed without merging**